Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp

```diff
@@ -597,9 +645,42 @@ static void webkit_media_stream_src_class_init(WebKitMediaStreamSrcClass* klass)
 
 static GstFlowReturn webkitMediaStreamSrcChain(GstPad* pad, GstObject* parent, GstBuffer* buffer)
 {
-    GRefPtr<GstElement> element = adoptGRef(GST_ELEMENT_CAST(gst_object_get_parent(parent)));
-    GstFlowReturn chainResult = gst_proxy_pad_chain_default(pad, GST_OBJECT_CAST(element.get()), buffer);
+    auto element = adoptGRef(GST_ELEMENT_CAST(gst_object_get_parent(parent)));
     auto* self = WEBKIT_MEDIA_STREAM_SRC_CAST(element.get());
+    GUniquePtr<char> name(gst_pad_get_name(pad));
+    auto padName = String::fromLatin1(name.get());
+
+    for (auto& source : self->priv->sources) {
+        if (source->padName() != padName)
+            continue;
+
+        Locker locker { *source->eosLocker() };
+        if (!source->eosPending())
+            continue;
+
+        // Make sure that the video.videoWidth is reset to 0.
+        webkitMediaStreamSrcPostStreamCollection(self);
+        auto tags = mediaStreamTrackPrivateGetTags(&source->track());
+        gst_pad_push_event(pad, gst_event_new_tag(tags.leakRef()));
+
+        auto* appSrc = source->get();
+        GST_STATE_LOCK(element.get());
+        gst_element_set_locked_state(appSrc, true);
+        gst_element_set_state(appSrc, GST_STATE_NULL);
+        gst_bin_remove(GST_BIN_CAST(self), appSrc);
+        gst_element_set_locked_state(appSrc, false);
+        GST_STATE_UNLOCK(element.get());
```

Comment:
Why don't you use a new scope and the automatic locker?

Reply:
Because lazy